runtime/lua/vim/lsp.lua

```diff
@@ -307,6 +307,260 @@ local function create_and_initialize_client(config)
+
+  for i = 1, select('#', ...) do
+    for k, v in
+      pairs(select(i, ...) or {} --[[@as table<string,any>]])
+    do
+      if ret[k] and client_config_list_keys[k] then
+        ret[k] = vim._ensure_list(ret[k])
+        vim.list_extend(ret[k], vim._ensure_list(v))
+      elseif type(ret[k]) == 'table' and type(v) == 'table' then
+        ret[k] = vim.tbl_deep_extend('force', ret[k], v)
+      else
+        ret[k] = v
+      end
+    end
+  end
+
+  return ret
+end
+
+--- @class vim.lsp.Config : vim.lsp.ClientConfig
+---
+--- See `cmd` in [vim.lsp.ClientConfig].
+--- @field cmd? string[]|fun(dispatchers: vim.lsp.rpc.Dispatchers): vim.lsp.rpc.PublicClient
+---
+--- Filetypes the client will attach to, if activated by `vim.lsp.enable()`.
+--- @field filetypes? string[]
+---
+--- Directory markers (.e.g. '.git/') where the LSP server will base its workspaceFolders,
+--- rootUri, and rootPath on initialization. Unused if `root_dir` is provided.
+--- @field root_markers? string[]
+---
+--- Predicate used to decide if a client should be re-used. Used on all
+--- running clients. The default implementation re-uses a client if name and
+--- root_dir matches.
+--- @field reuse_client? fun(client: vim.lsp.Client, config: vim.lsp.ClientConfig): boolean
+
+--- Update the configuration for an LSP client.
+---
+--- Use name '*' to set default configuration for all clients.
+---
+--- Can also be table assigned to redefine the configuration for a client.
+---
+--- Examples:
+---
+--- - Add a root marker for all clients:
+---   ```lua
+---     vim.lsp.config('*', {
+---         root_markers = { '.git' },
+---       })
+---       ```
+--- - Add additional capabilities to all clients:
+---   ```lua
+---     vim.lsp.config('*', {
+---         capabilities = {
+---           textDocument = {
+---             semanticTokens = {
+---               multilineTokenSupport = true,
+---             }
+---           }
+---         }
+---       })
+---   ```
+--- - (Re-)define the configuration for clangd:
+---   ```lua
+---     vim.lsp.config.clangd = {
+---         cmd = {
+---           'clangd',
+---           '--clang-tidy',
+---           '--background-index',
+---           '--offset-encoding=utf-8',
+---         },
+---         root_markers = { '.clangd', 'compile_commands.json' },
+---         filetypes = { 'c', 'cpp' },
+---       }
+---       ```
+--- - Get configuration for luals:
+---   ```lua
+---     local cfg = vim.lsp.config.luals
+---       ```
+---
+--- @param name string
+--- @param cfg vim.lsp.Config
+--- @diagnostic disable-next-line:assign-type-mismatch
+function lsp.config(name, cfg)
+  local _, _ = name, cfg -- ignore unused
+  -- dummy proto for docs
+end
+
+lsp._enabled_configs = {} --- @type table<string,{resolved_config:vim.lsp.Config?}>
+
+--- If a config in vim.lsp.config() is accessed then the resolved config becomes invalid.
+--- @param name string?
+local function invalidate_enabled_config(name)
+  if not name or name == '*' then
```

Comment:
Even though it's weird, there's nothing particular special about `''` as a name. Only `'*'` is considered special.

Reply:
> If we want to properly restrict this, then we would need to define exactly what valid names are allowed to be. Something like `[a-zA-Z0-9_-]+` would make sense since these names can be filenames in `lsp/*`.

Hmmm and that's probably also something to bring up in the specs.